src/FirebaseObject.cpp

```diff
@@ -16,57 +16,76 @@
 
 #include "FirebaseObject.h"
 
-namespace {
-template<typename T>
-T decodeJsonLiteral(const String& json) {
-  return JsonVariant{ArduinoJson::RawJson{json.c_str()}};
-}
-
-// ugly workaround to https://github.com/bblanchon/ArduinoJson/issues/265
-template<>
-String decodeJsonLiteral<String>(const String& json) {
-  StaticJsonBuffer<JSON_ARRAY_SIZE(1)> buf;
-  String array = "[" + json + "]";
-  return buf.parseArray(&array[0])[0];
-}
-}  // namespace
 
 FirebaseObject::FirebaseObject(const String& data) : data_{data} {
-  if (data_[0] == '{') {
-    json_ = &buffer_.parseObject(&data_[0]);
-  } else if (data_[0] == '"') {
-    data_ = decodeJsonLiteral<String>(data_);
-  }
+  json_ = buffer_.parse(&data_[0]);
+  // TODO(proppy): find a way to check decoding error, tricky because
+  // ArduinoJson doesn't surface error for variant parsing.
+  // See: https://github.com/bblanchon/ArduinoJson/issues/279
 }
 
-FirebaseObject::operator bool() {
-  return decodeJsonLiteral<bool>(data_);
+bool FirebaseObject::getBool(const String& path) {
+  JsonVariant variant = getJsonVariant(path);
+  if (!variant.is<bool>()) {
+    error_ = "failed to convert to bool";
+    return 0;
+  }
+  return static_cast<bool>(variant);
 }
 
-FirebaseObject::operator int() {
-    return decodeJsonLiteral<int>(data_);
+int FirebaseObject::getInt(const String& path) {
+  JsonVariant variant = getJsonVariant(path);
+  if (!variant.is<int>()) {
+    error_ = "failed to convert to int";
+    return 0;
+  }
+  return static_cast<int>(variant);
 }
 
-FirebaseObject::operator float() {
-    return decodeJsonLiteral<float>(data_);
+float FirebaseObject::getFloat(const String& path) {
+  JsonVariant variant = getJsonVariant(path);
+  if (!variant.is<float>()) {
+    error_ = "failed to convert to float";
+    return 0;
+  }
+  return static_cast<float>(variant);
 }
 
-FirebaseObject::operator const String&() {
-    return data_;
+String FirebaseObject::getString(const String& path) {
+  JsonVariant variant = getJsonVariant(path);
+  if (!variant.is<const char *>()) {
+    error_ = "failed to convert to string";
+    return "";
+  }
+  return static_cast<const char*>(variant);
 }
 
-FirebaseObject::operator const JsonObject&() {
-  return *json_;
+JsonVariant FirebaseObject::getJsonVariant(const String& path) {
+  String key(path);
+  char* start = &key[0];
+  char* end = start + key.length();
+  if (*start == '/') {
+    start++;
+  }
+  JsonVariant json = json_;
+  while (start < end) {
```

Comment:
While this block of code seems valid it is pretty tricky and could probably be made clearer, at the least it could use some comments. 
If I follow correctly you start with a pointer to the start of the path, you then search for the next / and replace it with a \0 you then read a c string from start that should then terminate on the recently added \0 to be the current token. You then start the next iteration immediately after the old / which is now a \0.